### PR TITLE
Add cucumber tests for recurring actions

### PR DESCRIPTION
**Pull request by wweellddeerr**
## What does this PR change?

Adjust recurring actions cucumber tests and add new scenarios for custom states

## GUI diff

No difference.

- [x] **DONE**

## Documentation
- No documentation needed: only internal and user invisible changes

- [x] **DONE**

## Test coverage
- Cucumber tests were added

- [x] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [x] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Review by Bischoff — CHANGES_REQUESTED** [review]
Thanks for this Welder

See my remarks here

https://github.com/SUSE/spacewalk/issues/21040#issuecomment-1573500529

**Review by ktsamis — APPROVED** [review]
To me it looks good but I still would like Eric to take a look. Thanks a lot for this Welder :)

File: testsuite/features/secondary/min_recurring_action.feature [hunk 1]
```diff
@@ -8,42 +8,114 @@ Feature: Recurring Actions
...
     Given I am authorized for the "Admin" section
 
-  Scenario: Create a minion Recurring Action
+  Scenario: Create a minion Custom state Recurring Action
+    When I am on the "Recurring Actions" page of this "sle_minion"
+    Then I should see a "No schedules created. Use Create to add a schedule" text
+    When I click on "Create"
+    Then I should see a "Action Type" text
+    When I select "Custom state" from "actionTypeDescription"
+    Then I wait until I see "Configure states to execute" text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:17]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 2]
```diff
@@ -8,42 +8,114 @@ Feature: Recurring Actions
...
+    And I check "Sync Modules-cbox"
+    And I click on "Save Changes"
+    Then I should see a "Edit State Ranks" text
+    And I should see a "Sync Modules" text
+    And I should not see a "Sync States" text
+    When I click on "Confirm"
+    Then I should see a "State assignments have been saved" text
+    And I enter "custom_state_schedule_name" as "scheduleName"
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:25]: When

File: testsuite/features/secondary/min_recurring_action.feature [hunk 3]
```diff
@@ -8,42 +8,114 @@ Feature: Recurring Actions
...
+    And I check radio button "schedule-daily"
+    And I enter 1 minutes from now as "time-daily_time"
+    And I click on "Create Schedule"
+    Then I wait until I see "Schedule successfully created" text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:29]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 4]
```diff
@@ -8,42 +8,114 @@ Feature: Recurring Actions
...
+    Then I should see a "SLS: util.syncstates" text
+    And I should not see a "SLS: util.syncmodules" text
+
+  Scenario: Cleanup: Delete a minion Custom state Recurring Action
+    When I am on the "Recurring Actions" page of this "sle_minion"
+    Then I should see a "custom_state_schedule_name_changed" text
+    When I click the "custom_state_schedule_name_changed" item delete button
+    Then I should see a "Delete Recurring Action Schedule" text
+    When I click on the red confirmation button
+    Then I wait until I see "Schedule 'custom_state_schedule_name_changed' has been deleted." text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:76]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 5]
```diff
@@ -8,42 +8,114 @@ Feature: Recurring Actions
...
     When I click on "Create"
     Then I should see a "Schedule Name" text
     When I enter "schedule_name" as "scheduleName"
     And I select "Highstate" from "actionTypeDescription"
     And I check radio button "schedule-daily"
-    And I pick "00:00" as time from "time-daily_time"
+    And I enter 1 minutes from now as "time-daily_time"
     And I click on the "disabled" toggler
     And I click on "Create Schedule"
     Then I wait until I see "Schedule successfully created" text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:89]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 6]
```diff
@@ -8,42 +8,114 @@ Feature: Recurring Actions
...
     When I am on the "Recurring Actions" page of this "sle_minion"
     Then I should see a "schedule_name" text
     When I click the "schedule_name" item edit button
     Then I should see a "Update Schedule" text
     When I enter "schedule_name_minion" as "scheduleName"
     And I check radio button "schedule-weekly"
+    And I select "Wednesday" from "date_weekly"
+    And I enter "01:35" as "time-weekly_time"
     And I click on "Update Schedule"
     Then I wait until I see "Schedule successfully updated" text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:107]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 7]
```diff
@@ -60,26 +137,22 @@ Feature: Recurring Actions
...
-    And I follow "Recurring-Action-test-group"
-    And I follow "Recurring Actions" in the content area
-    Then I should see a "schedule_name" text
-    When I click the "schedule_name" item details button
-    Then I should see a "Every day at 00:00" text
-    When I click on "Back"
-    Then I should see a "Schedules" text
+    When I am on the "Events" page of this "sle_minion"
+    And I follow "History"
+    Then I wait until I see the event "Apply recurring states [util.syncbeacons] scheduled by admin" completed during last minute, refreshing the page
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:155]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 8]
```diff
@@ -149,26 +245,23 @@ Feature: Recurring Actions
...
     And I check radio button "schedule-daily"
-    And I pick "00:00" as time from "time-daily_time"
+    And I enter 1 minutes from now as "time-daily_time"
     And I click on the "disabled" toggler
+    And I check "Hardware Profile Update-cbox"
+    And I click on "Save Changes"
+    Then I should see a "Edit State Ranks" text
+    And I click on "Confirm"
     And I click on "Create Schedule"
     Then I wait until I see "Schedule successfully created" text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:259]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 9]
```diff
@@ -149,26 +245,23 @@ Feature: Recurring Actions
...
-    And I follow "SUSE Test" in the content area
-    And I follow "Recurring Actions" in the content area
-    Then I should see a "schedule_name" text
-    When I click the "schedule_name" item details button
-    Then I should see a "Every day at 00:00" text
-    When I click on "Back"
-    Then I should see a "Schedules" text
+    When I am on the "Events" page of this "sle_minion"
+    And I follow "History"
+    Then I wait until I see the event "Apply recurring states [hardware.profileupdate] scheduled by admin" completed during last minute, refreshing the page
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:264]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 10]
```diff
@@ -195,34 +299,25 @@ Feature: Recurring Actions
...
+    When I follow the left menu "Admin > Organizations"
+    And I follow "SUSE Test" in the content area
+    And I follow "Recurring Actions" in the content area
+    Then I should see a "schedule_name_org" text
+    When I click the "schedule_name_org" item delete button
     Then I should see a "Delete Recurring Action Schedule" text
     When I click on the red confirmation button
-    Then I wait until I see "Schedule 'schedule_name_minion' has been deleted." text
-    Then I should see a "No schedules created. Use Create to add a schedule" text
+    Then I wait until I see "Schedule 'schedule_name_org' has been deleted." text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:318]: And

File: testsuite/features/secondary/min_recurring_action.feature [hunk 11]
```diff
@@ -233,7 +328,16 @@ Feature: Recurring Actions
...
     Then I should see a "No schedules created. Use Create to add a schedule" text
 
-  Scenario: Cleanup after running the tests
+  Scenario: Cleanup: Delete a minion Recurring Action
+    When I am on the "Recurring Actions" page of this "sle_minion"
+    Then I should see a "schedule_name_minion" text
+    When I click the "schedule_name_minion" item delete button
+    Then I should see a "Delete Recurring Action Schedule" text
+    When I click on the red confirmation button
+    Then I wait until I see "Schedule 'schedule_name_minion' has been deleted." text
```
Comment (Bischoff) [inline testsuite/features/secondary/min_recurring_action.feature:337]: And

**Review by Bischoff — CHANGES_REQUESTED** [review]
Thanks for the hard work, Welder.

Things I'd like to be changed:

1 - `Then` / `When` / `And` nitpicks (see details in this review)

2 - Some scenarios seem very long. Maybe it's possible to split them a bit?

3 - You now test that the events indeed appear as scheduled. That's a huge progress.
     But could we get further and test for real action results ? (file created, system setting changed, whatever)

4 - You are not testing the integration of custom actions (unless I missed it)


Thanks.

**Comment by jordimassaguerpla** [comment]
Hi! Sorry there was an issue with podman. The issue is fixed in master. Could you please rebase your PR? Sorry for any inconvenience.

**Review by Bischoff — APPROVED** [review]
Looking very good, thanks!

Just a last nitpick: you create *a* xyz, but then you edit/view/delete *the* xyz. Could you please fix that in the scenario titles?

**Pull request merged**